nncf/quantization/algorithms/weight_compression/lora.py

```diff
@@ -0,0 +1,208 @@
+from nncf.parameters import CompressWeightsMode
+from nncf.quantization.advanced_parameters import AdvancedLoraCorrectionParameters
+from nncf.quantization.algorithms.weight_compression.activation_stats import process_stats
+from nncf.quantization.algorithms.weight_compression.weight_lowering import do_dequantization
+from nncf.quantization.algorithms.weight_compression.weight_lowering import do_nf4_fake_quantization_from_norm_weight
+from nncf.tensor import Tensor
+from nncf.tensor import functions as fns
+from nncf.tensor.definitions import TensorDataType
+
+
+class DebugInterface:
+    def __init__(self):
+        self.debug_interface = {}
+
+    def add_loss(self, layer_name, value):
+        self.debug_interface[layer_name] = value
+
+    def dump_data(self):
+        if not self.debug_interface:
+            return
+        dump_dir = Path(DEBUG_LOG_DIR) / "lora"
+        dump_dir.mkdir(parents=True, exist_ok=True)
+
+        layer_dir = dump_dir / "per_layer"
+        layer_dir.mkdir(parents=True, exist_ok=True)
+
+        df = pd.DataFrame(self.debug_interface)
+        losses_path = dump_dir / "noises.csv"
+        nncf_logger.debug("Quantization noise through the correction process is saved to: ", losses_path)
+        df.to_csv(losses_path)
+
+        for name in df.columns:
+            plt.plot(df[name])
+            filename = name.replace("/", "_") + ".jpg"
+            plt.savefig(layer_dir / filename)
+            plt.clf()
+
+        delta = df.iloc[0] - df.iloc[-1]
+        nncf_logger.debug("Is quantization noise reduced for all layers: ", all(delta > 0))
+
+        _, ax = plt.subplots(1)
+        ax.plot(delta)
+        ax.set_xticklabels([])
+        delta_path = dump_dir / "qnoise_change.jpg"
+        nncf_logger.debug("Saving change in quantization noise for each layer to: ", delta_path)
+        plt.savefig(delta_path)
+        plt.clf()
+
+
+class LoraCorrectionAlgorithm:
+    def __init__(self, activations, lora_params: AdvancedLoraCorrectionParameters):
+        self._activations = activations
+        self._lora_correction_params = lora_params
+        self._debug_interface = DebugInterface() if is_debug() else None
+
+    @property
+    def is_int8_adapters(self) -> bool:
+        return self._lora_correction_params.is_int8_adapters
+
+    def is_applicable(self, wc_params):
+        return wc_params.compression_config.num_bits == 4
+
+    def calculate_adapters(self, weight, compressed_weight, wc_params):
+        try:
+            adapters = self.calculate_adapters_fn(
+                weight,
+                compressed_weight,
+                wc_params,
+                self._lora_correction_params,
+                self._activations,
+                self._debug_interface,
+            )
+        finally:
+            if self._debug_interface is not None:
+                self._debug_interface.dump_data()
+        return adapters
+
+    @staticmethod
+    def calculate_adapters_fn(
```

Comment:
The main functionality of the algorithm is here. @andreyanufr 
Designed to be called from other places, e.g. GPTQ @alexsu52